### PR TITLE
Add tests for OPTION calls #49

### PR DESCRIPTION
**Pull request by ChrBi**
This PR adds the test for the CORS-headers and the OPTION-calls.

### CODE REVIEW

File: tests/src/whisk/core/controller/test/RespondWithHeadersTests.scala [hunk 1]
```diff
@@ -44,47 +62,223 @@ class RespondWithHeadersTests extends ControllerTestCommon with RespondWithHeade
...
-                path("one") {
-                    complete(OK)
-                } ~ path("two") {
-                    complete(OK)
-                } ~ options {
-                    complete(OK)
-                } ~ reject
-            }
-        } ~ pathPrefix("other") {
-            complete(OK)
```
Comment (markusthoemmes) [inline tests/src/whisk/core/controller/test/RespondWithHeadersTests.scala]: These tests should've stayed as is. The new ones are no "Unit"-Tests but rather integration tests for the controller, as they need a deployed version of it. The new tests should go into a different package, `services` might be a good fit.

File: tests/src/services/HeadersTests.scala [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+import spray.http.HttpResponse
+import spray.http.StatusCodes.Accepted
+import spray.http.StatusCodes.OK
+import spray.http.Uri
+import spray.http.Uri.Path
+import spray.testkit.ScalatestRouteTest
+
+@RunWith(classOf[JUnitRunner])
+class HeadersTests extends FlatSpec
+    with ScalatestRouteTest
```
Comment (markusthoemmes) [inline tests/src/services/HeadersTests.scala]: Not needed

File: tests/src/services/HeadersTests.scala [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    val allowHeaders = `Access-Control-Allow-Headers`("Authorization", "Content-Type")
+
+    val url = Uri(s"http://${WhiskProperties.getControllerHost}:${WhiskProperties.getControllerPort}")
+    val pipeline: HttpRequest => Future[HttpResponse] = (
+        sendReceive
+        ~> unmarshal[HttpResponse])
+
+    val basePath = Path("/api/v1")
+
+    def toHttpMethod(method: String): Option[HttpMethod] = {
```
Comment (markusthoemmes) [inline tests/src/services/HeadersTests.scala]: Can use `HttpMethod.name` to extract the `String` from a `HttpMethod` to do the comparison.

File: tests/src/services/HeadersTests.scala [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+        } toMap
+
+        headers should contain allOf (allowOrigin, allowHeaders)
+        //        headersMap should contain key "Access-Control-Allow-Methods"
+        //
+        //        headersMap("Access-Control-Allow-Methods") should contain theSameElementsAs (allowedMethods)
+    }
+
+    it should "respond to OPTIONS with all headers" in {
+        val response = Await.result(pipeline(Options(url.withPath(basePath))), timeout)
```
Comment (markusthoemmes) [inline tests/src/services/HeadersTests.scala]: Can we use http://www.artima.com/docs-scalatest-2.0.M5/org/scalatest/concurrent/Futures.html to have more readable tests?

File: tests/src/services/HeadersTests.scala [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+            case "CONNECT" => Some(CONNECT)
+            case "DELETE"  => Some(DELETE)
+            case "GET"     => Some(GET)
+            case "HEAD"    => Some(HEAD)
+            case "OPTIONS" => Some(OPTIONS)
+            case "PATCH"   => Some(PATCH)
+            case "POST"    => Some(POST)
+            case "PUT"     => Some(PUT)
+            case "TRACE"   => Some(TRACE)
+            case x         => None
```
Comment (rabbah) [inline tests/src/services/HeadersTests.scala]: `x` -> `_`

File: docs/reference.md [hunk 6]
```diff
@@ -333,6 +333,11 @@ $ curl -u USERNAME:PASSWORD https://openwhisk.ng.bluemix.net/api/v1/namespaces/w
 ]
 ```
 
+The API of OpenWhisk also supports asynchronous calls on websites from the client.
+To allow these calls on the client side, we respond to OPTIONS calls with a list of allowed Headers and allowed Hosts.
```
Comment (markusthoemmes) [inline docs/reference.md]: To allow these calls from the browser directly, OpenWhisk responds to `OPTIONS` calls with  *Cross-Origin Resource Sharing* headers for allowed origins and headers.

File: docs/reference.md [hunk 7]
```diff
@@ -333,6 +333,11 @@ $ curl -u USERNAME:PASSWORD https://openwhisk.ng.bluemix.net/api/v1/namespaces/w
...
+The responses of the OpenWhisk-calls also include the header `Access-Control-Allow-Origin = *`
+
+**We do not recommend using this feature, because your authentication needs to be sent to OpenWhisk too. As result, you're credentials are on remote javascript clients and you do not have control over you're credentials anymore. An attacker would have the full control over your OpenWhisk account.**
```
Comment (markusthoemmes) [inline docs/reference.md]: **As OpenWhisk supports only one key per account currently, it is not recommended to use CORS beyond simple experiments currently. Your key would need to be embedded in a webapplication's code, thus being accessible by every viewer of that page. Use with caution.**

**Comment by markusthoemmes** [comment]
@csantanapr want to give this a pass documentationwise?

**Comment by csantanapr** [comment]
@markusthoemmes Changes to reference.md LGTM
@ChriBi always feel free to reach out directly to me 🙃

File: docs/reference.md [hunk 8]
```diff
@@ -333,6 +333,10 @@ $ curl -u USERNAME:PASSWORD https://openwhisk.ng.bluemix.net/api/v1/namespaces/w
 ]
 ```
 
+The API of OpenWhisk also supports asynchronous calls on websites from the client.
```
Comment (rabbah) [inline docs/reference.md]: -> The OpenWhisk API

File: docs/reference.md [hunk 9]
```diff
@@ -333,6 +333,10 @@ $ curl -u USERNAME:PASSWORD https://openwhisk.ng.bluemix.net/api/v1/namespaces/w
...
+To allow these calls from the browser directly, OpenWhisk responds to OPTIONS calls with Cross-Origin Resource Sharing headers for allowed origins and headers. Access-Control-Allow-Origin will always be * and Access-Control-Allow-Headers will always yield Authorization and Content-Type.
+
```
Comment (rabbah) [inline docs/reference.md]: Suggested rewording for the above:

The OpenWhisk API also supports request-response calls from web clients. OpenWhisk responds to `OPTIONS` requests with Cross-Origin Resource Sharing headers for allowed origins and headers. Currently, all origins are allowed (i.e., Access-Control-Allow-Origin  is `*`) and Access-Control-Allow-Headers yield Authorization and Content-Type.

File: docs/reference.md [hunk 10]
```diff
@@ -333,6 +333,10 @@ $ curl -u USERNAME:PASSWORD https://openwhisk.ng.bluemix.net/api/v1/namespaces/w
...
+**As OpenWhisk supports only one key per account currently, it is not recommended to use CORS beyond simple experiments at the moment. Your key would need to be embedded in a webapplication's code, thus being accessible by every viewer of that page. Use with caution.**
```
Comment (rabbah) [inline docs/reference.md]: ~~at the moment~~.

**Comment by rabbah** [comment]
suggest some tweaks to the doc prose. otherwise 👍

**Comment by markusthoemmes** [comment]
Included @rabbah's suggestions, rebased and merged in 5815204b148928dbfad6cafa9a26130ce5d9bd11

**Pull request closed without merging**